lib/HLSL/HLOperationLower.cpp

```diff
@@ -3586,7 +3586,89 @@ Value *GenerateRawBufLd(Value *handle, Value *bufIdx, Value *offset,
+    if (Constant *constProps = dyn_cast<Constant>(handleCall->getArgOperand(2))) {
+      RP = resource_helper::loadPropsFromConstant(*constProps);
+    }
+  }
+  return RP;
+}
+
+static Constant* GetRawBufLdStAlignmentInfo(IRBuilder<> &Builder, Value *handle, Type *EltTy,
+  Value *index, Value *offset, Constant *alignment) {
+
+  // If we don't have a constant offset for the element, then give up and
+  // return the default alignment -- i.e. element's allocation size in bytes.
+  if (!offset || !isa<ConstantInt>(offset)) {
+    return alignment;
+  }
+
+  CallInst *handleCall = dxilutil::FindCallToCreateHandle(handle);
+  if (!handleCall) {
+    return alignment;
+  }
+  hlsl::HLModule &HM = handleCall->getModule()->GetHLModule();
+  DxilResourceProperties RP = GetResourcePropertyFromHandleCall(&HM, handleCall);
+
+  // Bail out early if resource prop info not available
+  if (!RP.isValid()) {
+    return alignment;
+  }
+
+  // Only support structured buffer and byteaddress buffer for now
+  auto RK = RP.getResourceKind();
+  assert(DXIL::IsStructuredBuffer(RK) || DXIL::IsRawBuffer(RK));
+  if (!DXIL::IsStructuredBuffer(RK) && !DXIL::IsRawBuffer(RK)) {
+    return alignment;
+  }
+
+  constexpr unsigned dwordAlign = 4;
+  const unsigned alignmentVal = cast<ConstantInt>(alignment)->getLimitedValue();
+  unsigned offsetVal = cast<ConstantInt>(offset)->getLimitedValue();
+  if (DXIL::IsStructuredBuffer(RK)) {
+    unsigned structStrides = RP.StructStrideInBytes;
+    if (isa<ConstantInt>(index)) {
+      // If we have constant structure index, then it's possible to calculate
+      // the element address that can help figure out more accurate alignment
+      unsigned indexVal = cast<ConstantInt>(index)->getLimitedValue();
+      unsigned elemAddr = (indexVal * structStrides) + offsetVal;
+      if (elemAddr % dwordAlign == 0 && alignmentVal < dwordAlign) {
+        return ConstantInt::get(Builder.getInt32Ty(), dwordAlign);
+      }
+    }
+    else {
+      // Even if only offset is constant, we can still do a better job
+      // than defaulting to element's size (in bytes).
+      if (structStrides % dwordAlign == 0 && offsetVal % dwordAlign == 0 && alignmentVal < dwordAlign) {
+        return ConstantInt::get(Builder.getInt32Ty(), dwordAlign);
+      }
+    }
+  }
+  else if (DXIL::IsRawBuffer(RK)) {
+    if (offsetVal % dwordAlign == 0 && alignmentVal < dwordAlign) {
+      return ConstantInt::get(Builder.getInt32Ty(), dwordAlign);
+    }
+  }
+  return alignment;
+}
+
+static unsigned GetRawBufLdStAlignmentInfo(IRBuilder<> &Builder, Value *handle, const DataLayout &DL,
+  Type *EltTy, Value *index, Value *offset) {
+
+  // Use allocation size as the default alignment
+  Constant *alignment = ConstantInt::get(Builder.getInt32Ty(), DL.getTypeAllocSize(EltTy));
+  alignment = GetRawBufLdStAlignmentInfo(Builder, handle, EltTy, index, offset, alignment);
+  return cast<ConstantInt>(alignment)->getLimitedValue();
+}
```

Comment:
This is the key function which attempts to calculate the more accurate alignment information.